vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,414 @@
+			RequestMetadata:    requestMetadata,
+		}
+		generateDEKResponse, err := kmsCryptoClient.GenerateDataEncryptionKey(context.Background(), dekInput)
+		if err != nil || generateDEKResponse.Ciphertext == nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("failed keyID validation: {{err}}", err)
+		}
+		k.logger.Info("successfully validated keyID")
+
+		// Store client
+		k.cryptoClient = kmsCryptoClient
+	}
+
+	// Check and set OCI KMS management client
+	if k.managementClient == nil {
+		kmsManagementClient, err := k.getOCIKMSManagementClient()
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("error initializing OCI KMS client: {{err}}", err)
+		}
+
+		keyVersion, err := k.getCurrentKeyVersion(kmsManagementClient)
+		if err != nil {
+			metrics.IncrCounter(metricInitFailed, 1)
+			return nil, errwrap.Wrapf("error initializing OCI KMS client: {{err}}", err)
+		}
+		k.currentKeyID.Store(keyVersion)
+
+		k.managementClient = kmsManagementClient
+	}
+
+	// Map that holds non-sensitive configuration info
+	sealInfo := make(map[string]string)
+	sealInfo[KMSConfigKeyID] = k.keyID
+	sealInfo[KMSConfigCryptoEndpoint] = k.cryptoEndpoint
+	sealInfo[KMSConfigManagementEndpoint] = k.managementEndpoint
+
+	return sealInfo, nil
+}
+
+func (k *OCIKMSSeal) SealType() string {
+	return seal.OCIKMS
+}
+
+func (k *OCIKMSSeal) KeyID() string {
+	return k.currentKeyID.Load().(string)
+}
+
+func (k *OCIKMSSeal) Init(context.Context) error {
+	return nil
+}
+
+func (k *OCIKMSSeal) Finalize(context.Context) error {
+	return nil
+}
+
+func (k *OCIKMSSeal) Encrypt(ctx context.Context, plaintext []byte) (*physical.EncryptedBlobInfo, error) {
+	defer metrics.MeasureSince(metricEncrypt, time.Now())
+	if plaintext == nil || len(plaintext) == 0 {
+		metrics.IncrCounter(metricEncryptFailed, 1)
+		return nil, fmt.Errorf("given plaintext for encryption is nil")
+	}
+
+	env, err := seal.NewEnvelope().Encrypt(plaintext)
+	if err != nil {
+		metrics.IncrCounter(metricEncryptFailed, 1)
+		return nil, errwrap.Wrapf("error wrapping data: {{err}}", err)
+	}
+
+	if k.cryptoClient == nil {
+		metrics.IncrCounter(metricEncryptFailed, 1)
+		return nil, fmt.Errorf("nil client")
+	}
+
+	// OCI KMS required base64 encrypted plain text before sending to the service
+	encodedKey := base64.StdEncoding.EncodeToString(env.Key)
+
+	// Build Encrypt Request
+	requestMetadata := k.getRequestMetadata()
+	encryptedDataDetails := keymanagement.EncryptDataDetails{
+		KeyId:     &k.keyID,
+		Plaintext: &encodedKey,
+	}
+
+	input := keymanagement.EncryptRequest{
+		EncryptDataDetails: encryptedDataDetails,
+		RequestMetadata:    requestMetadata,
+	}
+	output, err := k.cryptoClient.Encrypt(ctx, input)
+	if err != nil {
+		metrics.IncrCounter(metricEncryptFailed, 1)
+		return nil, errwrap.Wrapf("error encrypting data: {{err}}", err)
+	}
+
+	keyVersion, err := k.getCurrentKeyVersion(k.managementClient)
```

Comment:
It may be worth noting in a comment that there is a potential timing issue if the key gets rotated between this operation and the operation above.